src/main/java/use_case/notification/NotificationObserver.java

```diff
@@ -0,0 +1,117 @@
+package main.java.use_case.notification;
+
+import main.java.constants.NotificationType;
+import main.java.entity.Notification;
+import main.java.entity_gateway.CalendarManager;
+import main.java.entity_gateway.EventReader;
+import main.java.entity_gateway.TaskReader;
+import main.java.entity_gateway.TodoListManager;
+
+import java.time.Duration;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.util.Set;
+
+/**
+ * NotificationObserver is responsible for receiving and
+ * processing notification command from NotificationTracker.
+ * NotificationObserver will organize the message format based
+ * on the NotificationType of notifications.
+ *
+ * - sender: responsible for sending the message text to the user on several platforms
+ * - calendarManager: used for searching events based on event ids
+ * - todoListManager: used for searching tasks based on task ids
+ */
+public class NotificationObserver implements INotificationObserver {
+
+    private final NotificationSender sender;
+    private final CalendarManager calendarManager;
+    private final TodoListManager todoListManager;
+
+    public NotificationObserver(NotificationSender sender, CalendarManager calendarManager,
+                                TodoListManager todoListManager) {
+        this.sender = sender;
+        this.calendarManager = calendarManager;
+        this.todoListManager = todoListManager;
+    }
+
+    /**
+     * Receive new notification command and send it out
+     * @param notification the new notification that will be sent out
+     * @return true if the message is sent out successfully
+     */
+    @Override
+    public boolean update(Notification notification) {
+        if (notification.getNotificationType() == NotificationType.Task) {
+            int taskId = notification.getAssociatedId();
+            String message = taskMessage(taskId);
+            return sender.sendMessage(message);
+        } else if (notification.getNotificationType() == NotificationType.Event) {
+            int eventId = notification.getAssociatedId();
+            String message = eventMessage(eventId);
+            return sender.sendMessage(message);
+        }
+        return false;
+    }
+
+    /**
+     * Search for information of a task using the task id,
+     * and generate message text for this task:
+     *
+     * "You have an upcoming event <taskName> due on <dueDate> at <dueTime>."
+     *
+     * @param taskId the id of the task
+     * @return string of message text for the task
+     */
+    private String taskMessage(int taskId) {
+        TaskReader taskReader = todoListManager.getTask(taskId);
+        String taskName = taskReader.getName();
+        LocalDate dueDate = taskReader.getDeadline().toLocalDate();
+        LocalTime dueTime = taskReader.getDeadline().toLocalTime();
+        return "You have a task " + taskName + " due on " + dueDate + " at " + dueTime + ".";
```

Comment:
Should the text be generated by a use case?

Reply:
That is a good point oops. At first Junru and I were thinking that since NotificationObserver is the one that differentiates "event" and "task" notification, might as well make it create the message that is specific to the type of notification. 

Should we move it to a presenter class instead?